### PR TITLE
feat: add context object to log messages

### PR DESCRIPTION
**Pull request by raman325**
I'm not sure that everything I've captured is strictly necessary so we can relax some of this based on feedback

### CODE REVIEW

File: packages/zwave-js/src/lib/log/Driver.ts [hunk 1]
```diff
@@ -53,6 +54,7 @@ export class DriverLogger extends ZWaveLoggerBase {
 			level: actualLevel,
 			message,
 			direction: getDirectionPrefix("none"),
+			context: CONTEXT,
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/log/Driver.ts]:
```suggestion
			context: { ...CONTEXT },
```
would avoid a situation where applications can change our `CONTEXT` object's contents.

File: packages/zwave-js/src/lib/log/Driver.ts [hunk 2]
```diff
@@ -219,6 +222,7 @@ export class DriverLogger extends ZWaveLoggerBase {
 				queue.length === 1 ? "" : "s"
 			})`,
 			direction: getDirectionPrefix("none"),
+			context: CONTEXT,
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/log/Driver.ts]: dito

**Review by AlCalzone — CHANGES_REQUESTED** [review]
I'm wondering if `context` should have a `type` property or something that can be used to distinguish between several possible contexts an application might encounter.

Also, for type safety, we should probably define the different specific context types and export them.

**Comment by raman325** [comment]
> I'm wondering if `context` should have a `type` property or something that can be used to distinguish between several possible contexts an application might encounter.
> 
> Also, for type safety, we should probably define the different specific context types and export them.

That's `source` the way I have currently implemented it, but we can switch to `type` since you use that elsewhere. And stronger typing is a good idea

**Comment by raman325** [comment]
@AlCalzone when you say you want to type the different context types, do you mean for example typing https://github.com/zwave-js/node-zwave-js/pull/3251/files#diff-7076042951692d76ef597b7bae4634cb8110f9a95eb8708065fdf41cd0a4c919R98-R103 and https://github.com/zwave-js/node-zwave-js/pull/3251/files#diff-7076042951692d76ef597b7bae4634cb8110f9a95eb8708065fdf41cd0a4c919R144-R151 as distinct types?

**Comment by AlCalzone** [comment]
> That's `source` the way I have currently implemented it

Yes, but it doesn't distinguish between the possible contexts, e.g. there are multiple `source: "controller"` contexts.


> @AlCalzone when you say you want to type the different context types, do you mean for example typing [#3251 (files)](https://github.com/zwave-js/node-zwave-js/pull/3251/files#diff-7076042951692d76ef597b7bae4634cb8110f9a95eb8708065fdf41cd0a4c919R98-R103) and [#3251 (files)](https://github.com/zwave-js/node-zwave-js/pull/3251/files#diff-7076042951692d76ef597b7bae4634cb8110f9a95eb8708065fdf41cd0a4c919R144-R151) as distinct types?

Yep

**Comment by raman325** [comment]
@zwave-js-bot fix lint

**Comment by raman325** [comment]
@zwave-js-bot fix lint

**Comment by AlCalzone** [comment]
@raman325 let me know when you're done here :)

**Comment by raman325** [comment]
I'm done, this is ready for review!

**Comment by AlCalzone** [comment]
Sorry it took me so long. I've made a few changes:
* exported all log context types
* converted some of the types to generics to let TypeScript validate the contexts better
* made the `type` property to distinguish logs from the same logger optional
* removed the limitation of `source`, so applications which hook into the logging functionality (like Node Red) can still do it.

**Review by AlCalzone — APPROVED** [review]
OK with my changes - @raman325 can you give this another test?

**Comment by raman325** [comment]
> OK with my changes - @raman325 can you give this another test?

I will test this in the next day or two and get back to you

**Comment by AlCalzone** [comment]
Huh?

**Comment by raman325** [comment]
uhh this was not intentional. Shit

**Comment by raman325** [comment]
sorry for the delay @AlCalzone but I tested the changes and everything looks good! I think we are OK to merge

**Pull request merged**